### PR TITLE
TEPHRA-80 Support configurable conflict detection in TransactionAwareHTable

### PR DESCRIPTION
**Pull request by ghelmling**

### CODE REVIEW

File: tephra-core/src/main/java/co/cask/tephra/TxConstants.java [hunk 1]
```diff
@@ -26,7 +26,23 @@
  */
 public class TxConstants {
   /**
-   * property set for {@link org.apache.hadoop.hbase.HColumnDescriptor} to configure time-to-live on data within
+   * Defines what level of conflict detection should be used for transactions.  {@code ROW} means that only the
+   * table name and row key for each change will be used to determine if the transaction change sets conflict.
+   * {@code ROWCOL} means that the table name, row key, column family, and column qualifier will all be used to
```
Comment (abaranau) [inline tephra-core/src/main/java/co/cask/tephra/TxConstants.java]: COLUMN - not ROWCOL

File: tephra-core/src/main/java/co/cask/tephra/TxConstants.java [hunk 2]
```diff
@@ -26,7 +26,23 @@
...
+   * identify write conflicts.
+   *
+   * <p>
+   * <strong>Note: for a given table, all clients must use the same conflict detection setting!</strong>
+   * Otherwise conflicts will not be flagged correctly.
+   * </p>
+   */
+  public enum ConflictDetection {
+    ROW,
+    COLUMN
```
Comment (abaranau) [inline tephra-core/src/main/java/co/cask/tephra/TxConstants.java]: how about NONE?

File: tephra-hbase-compat-0.94/src/main/java/co/cask/tephra/hbase94/TransactionAwareHTable.java [hunk 3]
```diff
@@ -59,27 +60,54 @@
...
    * Create a transactional aware instance of the passed HTable, with the option
    * of allowing non-transactional operations.
-   * @param hTable
-   * @param allowNonTransactional
+   * @param hTable underlying HBase table to use
+   * @param allowNonTransactional if true, additional operations (checkAndPut, increment, checkAndDelete)
+   *                              will be available, though non-transactional
    */
   public TransactionAwareHTable(HTableInterface hTable, boolean allowNonTransactional) {
+    this(hTable, TxConstants.ConflictDetection.COLUMN, allowNonTransactional);
```
Comment (abaranau) [inline tephra-hbase-compat-0.94/src/main/java/co/cask/tephra/hbase94/TransactionAwareHTable.java]: do we say anywhere explicitly that column is default? or do we need to say that

File: tephra-hbase-compat-0.94/src/main/java/co/cask/tephra/hbase94/TransactionAwareHTable.java [hunk 4]
```diff
@@ -380,10 +408,24 @@ public void startTx(Transaction tx) {
...
+        }
+        break;
+      case COLUMN:
+        txChanges = new ArrayList<byte[]>(changeSet.size());
+        for (ActionChange change : changeSet) {
+          txChanges.add(Bytes.add(getTableName(), change.getRow(),
+              Bytes.add(change.getFamily(), change.getQualifier())));
+        }
+        break;
+      default:
```
Comment (abaranau) [inline tephra-hbase-compat-0.94/src/main/java/co/cask/tephra/hbase94/TransactionAwareHTable.java]: should be very easy to add NONE ;)

**Comment by abaranau** [comment]
couple comments

**Comment by abaranau** [comment]
alright, after discussion, we'll figure out NONE later. Looks good to merge then

**Comment by ghelmling** [comment]
Fixed TxConstants doc comments and added a note that COLUMN is the default.

**Comment by abaranau** [comment]
LGTM

**Pull request merged**